### PR TITLE
Undo redo

### PR DESCRIPTION
**Pull request by RichardNemcovic**
Resolved failing tests of other branches in order to be able to build the project again and pass the pipeline tests.

### CODE REVIEW

File: .github/workflows/build_verify_pipeline.yml [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+      - name: Check out repository
+        uses: actions/checkout@v2
+      - name: JDK setup, Version 1.8
+        uses: actions/setup-java@v1
+        with:
+          java-version: 1.8
+      - name: Clean and build
+        run: mvn -B clean compile --file JHotDraw/pom.xml
+      - name: Run tests
+        run: mvn -B test --file JHotDraw/pom.xml
```
Comment (do-nat) [inline .github/workflows/build_verify_pipeline.yml:30]: there is already a pipeline file and this file should not be duplicated in the master branch

File: .github/workflows/build_verify_pipeline.yml [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+      - refactorings
+      - workflow
+      - master
+      - UndoRedo
+  pull_request:
+    branches:
+      - refactorings
+      - workflow
+      - master
+      - UndoRedo
```
Comment (do-nat) [inline .github/workflows/build_verify_pipeline.yml:15]: there's no use in limiting the builds on branches here. You should maybe look up what this implicated :)

File: JHotDraw/src/test/java/org/jhotdraw/undo/UndoRedoManagerTest.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    }
+
+    @Test
+    public void undoOrRedoTest_other(){
+        try{
+            undoRedoManager.undoOrRedo(UndoRedoManager.UndoRedoAction.TEST);
+            fail();
+        } catch (InvalidParameterException e){
+            System.out.println(e.getMessage());
+        }
```
Comment (do-nat) [inline JHotDraw/src/test/java/org/jhotdraw/undo/UndoRedoManagerTest.java:52]: what does this do? does this make sense?

File: JHotDraw/src/test/java/org/jhotdraw/undo/UndoRedoManagerTest.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+    @Before
+    public void setUp() {
+        undoRedoManager = new UndoRedoManager();
+    }
+
+    @Test
+    public void undoOrRedoTest_UNDO(){
+        UndoRedoManager.UndoRedoAction action = undoRedoManager.undoOrRedo(UndoRedoManager.UndoRedoAction.UNDO);
+        assertEquals(action, UndoRedoManager.UndoRedoAction.UNDO);
```
Comment (do-nat) [inline JHotDraw/src/test/java/org/jhotdraw/undo/UndoRedoManagerTest.java:24]: these tests seem to test the return value of the method undoOrRedo(). Do they actually do more than that?

File: JHotDraw/src/test/java/org/jhotdraw/samples/svg/gui/CanvasToolBarTest.java [hunk 5]
```diff
@@ -1,6 +1,6 @@
 package org.jhotdraw.samples.svg.gui;
```
Comment (do-nat) [inline JHotDraw/src/test/java/org/jhotdraw/samples/svg/gui/CanvasToolBarTest.java:1]: this test works in master, you don't have to adjust it (revert these changes!)

File: JHotDraw/src/test/java/org/jhotdraw/samples/svg/figures/EllipseToolDrawAcceptanceTest.java [hunk 6]
```diff
@@ -5,6 +5,7 @@
  */
```
Comment (do-nat) [inline JHotDraw/src/test/java/org/jhotdraw/samples/svg/figures/EllipseToolDrawAcceptanceTest.java:5]: this test works on master, you can revert your changes

File: JHotDraw/src/test/java/org/jhotdraw/draw/action/SendToBackActionTest.java [hunk 7]
```diff
@@ -28,10 +28,12 @@ public void testGetID() {
     @Test
```
Comment (do-nat) [inline JHotDraw/src/test/java/org/jhotdraw/draw/action/SendToBackActionTest.java:28]: this test works on master, you can revert your changes

File: JHotDraw/src/test/java/org/jhotdraw/draw/action/BringToFrontActionTest.java [hunk 8]
```diff
@@ -23,16 +23,16 @@ public void testGetID() {
         String expResult = "edit.bringToFront";
```
Comment (do-nat) [inline JHotDraw/src/test/java/org/jhotdraw/draw/action/BringToFrontActionTest.java:23]: this test works on master, you can revert your changes

File: JHotDraw/src/test/java/org/jhotdraw/draw/WhenUndoActionIsPerformed.java [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+    private void prepare() {
+        map = new HashMap<>();
+        mouseEvent = mock(MouseEvent.class);
+        creationTool = new CreationTool(argumentList.svgBezierFigure, map);
+        creationTool.createdFigure = argumentList.svgBezierFigure;
+        creationTool.editor = argumentList.editor;
+    }
+
+    public WhenUndoActionIsPerformed undo_action_is_performed() {
+        List <Figure> list = argumentList.svgBezierFigure.getDrawing().getChildren();
```
Comment (do-nat) [inline JHotDraw/src/test/java/org/jhotdraw/draw/WhenUndoActionIsPerformed.java:31]: you don't use this anymore

File: JHotDraw/src/test/java/org/jhotdraw/draw/WhenUndoActionIsPerformed.java [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+import java.awt.event.*;
+import java.util.*;
+
+import static org.mockito.Mockito.mock;
+
+public class WhenUndoActionIsPerformed extends Stage<WhenUndoActionIsPerformed> {
+    @ExpectedScenarioState
+    protected ArgumentList argumentList;
+
+    private HashMap<AttributeKey, Object> map;
```
Comment (do-nat) [inline JHotDraw/src/test/java/org/jhotdraw/draw/WhenUndoActionIsPerformed.java:17]: this does not have to be a property, can be a simple variable

File: JHotDraw/src/main/java/org/jhotdraw/undo/UndoRedoManager.java [hunk 11]
```diff
@@ -243,58 +252,71 @@ private void updateActions() {
...
     }
-    
+
     public void addPropertyChangeListener(PropertyChangeListener listener) {
         propertySupport.addPropertyChangeListener(listener);
     }
+
+    //TODO remove code smell (Speculative Generality)
+    /*
     public void addPropertyChangeListener(String propertyName, PropertyChangeListener listener) {
```
Comment (do-nat) [inline JHotDraw/src/main/java/org/jhotdraw/undo/UndoRedoManager.java:320]: remove commented code

File: JHotDraw/src/main/java/org/jhotdraw/undo/UndoRedoManager.java [hunk 12]
```diff
@@ -155,21 +155,32 @@ public void discardAllEdits() {
...
+
+
+
+
+
+
+
+
+
+
```
Comment (do-nat) [inline JHotDraw/src/main/java/org/jhotdraw/undo/UndoRedoManager.java:168]: these can be removed (also, these formatting changes in this file should be avoided!)

File: JHotDraw/src/main/java/org/jhotdraw/undo/UndoRedoManager.java [hunk 13]
```diff
@@ -155,21 +155,32 @@ public void discardAllEdits() {
...
     public void setHasSignificantEdits(boolean newValue) {
         boolean oldValue = hasSignificantEdits;
         hasSignificantEdits = newValue;
-        firePropertyChange("hasSignificantEdits", oldValue, newValue);
+        //firePropertyChange("hasSignificantEdits", oldValue, newValue);
```
Comment (do-nat) [inline JHotDraw/src/main/java/org/jhotdraw/undo/UndoRedoManager.java:172]: remove line

**Review by do-nat — CHANGES_REQUESTED** [review]
there's still some things to be done :blush:

File: .github/workflows/build_verify_pipeline.yml (hunk 1 above)
Comment (RichardNemcovic) [reply .github/workflows/build_verify_pipeline.yml:30]: Thanks for the information, I removed the other one.

File: JHotDraw/src/test/java/org/jhotdraw/undo/UndoRedoManagerTest.java (hunk 4 above)
Comment (RichardNemcovic) [reply JHotDraw/src/test/java/org/jhotdraw/undo/UndoRedoManagerTest.java:24]: These test check, if there was an appropriate workflow, based on precondtions

File: JHotDraw/src/test/java/org/jhotdraw/samples/svg/gui/CanvasToolBarTest.java (hunk 5 above)
Comment (RichardNemcovic) [reply JHotDraw/src/test/java/org/jhotdraw/samples/svg/gui/CanvasToolBarTest.java:1]: Done

File: JHotDraw/src/test/java/org/jhotdraw/samples/svg/figures/EllipseToolDrawAcceptanceTest.java (hunk 6 above)
Comment (RichardNemcovic) [reply JHotDraw/src/test/java/org/jhotdraw/samples/svg/figures/EllipseToolDrawAcceptanceTest.java:5]: done

File: JHotDraw/src/test/java/org/jhotdraw/draw/action/SendToBackActionTest.java (hunk 7 above)
Comment (RichardNemcovic) [reply JHotDraw/src/test/java/org/jhotdraw/draw/action/SendToBackActionTest.java:28]: done

File: JHotDraw/src/test/java/org/jhotdraw/draw/action/BringToFrontActionTest.java (hunk 8 above)
Comment (RichardNemcovic) [reply JHotDraw/src/test/java/org/jhotdraw/draw/action/BringToFrontActionTest.java:23]: done

File: JHotDraw/src/test/java/org/jhotdraw/draw/WhenUndoActionIsPerformed.java (hunk 9 above)
Comment (RichardNemcovic) [reply JHotDraw/src/test/java/org/jhotdraw/draw/WhenUndoActionIsPerformed.java:31]: deleted

File: JHotDraw/src/test/java/org/jhotdraw/draw/WhenUndoActionIsPerformed.java (hunk 10 above)
Comment (RichardNemcovic) [reply JHotDraw/src/test/java/org/jhotdraw/draw/WhenUndoActionIsPerformed.java:17]: True, done

File: JHotDraw/src/main/java/org/jhotdraw/undo/UndoRedoManager.java (hunk 11 above)
Comment (RichardNemcovic) [reply JHotDraw/src/main/java/org/jhotdraw/undo/UndoRedoManager.java:320]: done

File: JHotDraw/src/main/java/org/jhotdraw/undo/UndoRedoManager.java (hunk 12 above)
Comment (RichardNemcovic) [reply JHotDraw/src/main/java/org/jhotdraw/undo/UndoRedoManager.java:168]: done

File: JHotDraw/src/main/java/org/jhotdraw/undo/UndoRedoManager.java (hunk 13 above)
Comment (RichardNemcovic) [reply JHotDraw/src/main/java/org/jhotdraw/undo/UndoRedoManager.java:172]: done

File: .github/workflows/build_verify_pipeline.yml (hunk 1 above)
Comment (RichardNemcovic) [reply .github/workflows/build_verify_pipeline.yml:30]: done

File: JHotDraw/src/test/java/org/jhotdraw/undo/UndoRedoManagerTest.java (hunk 4 above)
Comment (RichardNemcovic) [reply JHotDraw/src/test/java/org/jhotdraw/undo/UndoRedoManagerTest.java:24]: They check the correctness of the workflow, but apart from that no.

File: JHotDraw/src/test/java/org/jhotdraw/samples/svg/gui/CanvasToolBarTest.java (hunk 5 above)
Comment (RichardNemcovic) [reply JHotDraw/src/test/java/org/jhotdraw/samples/svg/gui/CanvasToolBarTest.java:1]: done

File: JHotDraw/src/test/java/org/jhotdraw/undo/UndoRedoManagerTest.java (hunk 3 above)
Comment (RichardNemcovic) [reply JHotDraw/src/test/java/org/jhotdraw/undo/UndoRedoManagerTest.java:52]: Yes, the "undoRedoManager.undoOrRedo(UndoRedoManager.UndoRedoAction.TEST);" should throw an error and if not, the test should fail. If it throws, it just prints out the error message.

**Review by do-nat — APPROVED** [review]

**Pull request merged**